pkg/virt-handler/cgroup/util.go

```diff
@@ -0,0 +1,89 @@
+package cgroup
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"syscall"
+
+	runc_configs "github.com/opencontainers/runc/libcontainer/configs"
+
+	"kubevirt.io/client-go/log"
+)
+
+const (
+	cgroupStr = "cgroup"
+
+	procMountPoint = "/proc"
+
+	HostRootPath       = "/proc/1/root"
+	cgroupBasePath     = "/sys/fs/" + cgroupStr
+	HostCgroupBasePath = HostRootPath + cgroupBasePath
+)
+
+// Templates for logging / error messages
+const (
+	errApplyingDeviceRule = "error occurred while applying device rule: %v"
+	errApplyingOtherRules = "error occurred while applying rules (that are not device rules): %v"
+	settingDeviceRule     = "setting device rule for cgroup %s: %v"
+
+	v1Str = "v1"
+	v2Str = "v2"
+
+	loggingVerbosity = 2
+)
+
+// getNewResourcesWithoutDevices returns a new Resources struct with Devices attributes dropped
+func getNewResourcesWithoutDevices(r *runc_configs.Resources) runc_configs.Resources {
+	resourcesWithoutDevices := *r
+	resourcesWithoutDevices.Devices = nil
+
+	return resourcesWithoutDevices
+}
+
+func logAndReturnErrorWithSprintfIfNotNil(err error, template string, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+
+	newErr := fmt.Errorf(template, args...)
+	log.Log.Error(newErr.Error())
+	return newErr
+}
+
+func areResourcesEmpty(r runc_configs.Resources) bool {
+	return reflect.DeepEqual(r, runc_configs.Resources{})
+}
+
+// RunWithChroot changes the root directory (via "chroot") into newPath, then
+// runs toRun function. When the function finishes, changes back the root directory
+// to the original one that
+func RunWithChroot(newPath string, toRun func() error) error { // ihol3 bad place to define func
+	originalRoot, err := os.Open("/")
+	if err != nil {
+		return fmt.Errorf("failed to run with chroot - failed to open root directory. error: %v", err)
+	}
+
+	err = syscall.Chroot(newPath)
```

Comment:
One more thought that comes to my mind: is it safe to `chroot` here? It will affect the whole process (`virt-handler` I guess) and other threads may be doing smth with relative filepath meanwhile.

Reply:
AFAIK syscall.Chroot affects the entire process, not just a single thread.

Reply:
if you do something like this, you must ensure that the go routine is locked (`LockOSThread`) and you have to ensure that a single go routine runs on that thread. If you can revert you can the use `UnlockOSThread` (relatively cheap but locks a whole thead). Otherwise you have to just exit the go routine to kill the whole thread with it (expensive since a whole new thread needs to be started, including context copy).

Reply:
Thanks @rmohr!

Reply:
Thank you @vasiliy-ul for catching that!
@rmohr As I wrote a while back:

> The reason I was doing this is because runc (for some reason) uses absolute path in which the root is cgroups root. So I'm chrooting there, letting runc operate, then coming back.

Do you have any idea how we can overcome this? Doesn't Go has a chroot alternative for running one goroutine?